mountpoint-s3/src/inode.rs

```diff
@@ -518,14 +545,60 @@ impl SuperblockInner {
             return Err(InodeError::InvalidFileName(name.into()));
         }
 
-        // TODO use caches. if we already know about this name, we just need to revalidate the stat
-        // cache and then read it.
-        let remote = self.remote_lookup(client, parent_ino, name).await?;
-        let lookup = self.update_from_remote(parent_ino, name, remote)?;
+        let lookup = if skip_cache {
+            None
+        } else {
+            self.cache_lookup(parent_ino, name)
+        };
+
+        let lookup = match lookup {
+            Some(lookup) => lookup,
+            None => {
+                let remote = self.remote_lookup(client, parent_ino, name).await?;
+                self.update_from_remote(parent_ino, name, remote)?
+            }
+        };
+
         lookup.inode.verify_child(parent_ino, name.as_ref())?;
         Ok(lookup)
     }
 
+    /// Lookup an [Inode] against known directory entries in the parent,
+    /// verifying any returned entry has not expired.
+    fn cache_lookup(&self, parent_ino: InodeNo, name: &str) -> Option<LookedUp> {
+        fn do_cache_lookup(parent: Inode, name: &str) -> Option<LookedUp> {
```

Comment:
nit: I'm not sure the local fn is appropriate here. Is there no way to make this logic more linear?

Reply:
I'm wanting to make the core logic clear while still letting us wrap it in traces and metrics.

The difficulty is that there's a lot of `Option` and `Result` unpacking, and so it gets pretty hard to follow if trying to eliminate the use of the `?` operator.

I gave it a go and played around for a bit, but I'm not happy with it:

```rust


    /// Lookup an [Inode] against known directory entries in the parent,
    /// verifying any returned entry has not expired.
    fn cache_lookup(&self, parent_ino: InodeNo, name: &str) -> Option<LookedUp> {
        let parent = self.get(parent_ino).ok()?;
        let lookup = match &parent.get_inode_state().ok()?.kind_data {
            InodeKindData::File { .. } => unreachable!("parent should be a directory!"),
            InodeKindData::Directory { children, .. } => {
                if let Some(inode) = children.get(name) {
                    let valid_stat = inode.get_inode_state().ok()
                        .map(|state| state.stat)
                        .filter(|stat| stat.is_valid());
                    if let Some(valid_stat) = valid_stat {
                        let lookup = LookedUp {
                            inode: inode.clone(),
                            stat: valid_stat.clone(),
                        };
                        Some(lookup)
                    } else {
                        None
                    }
                } else {
                    None
                }
            }
        };

        match &lookup {
            Some(lookup) => trace!("lookup returned from cache: {:?}", lookup),
            None => trace!("no lookup available from cache"),
        }
        metrics::counter!("metadata_cache.cache_hit", lookup.is_some().into());

        lookup
    }
```

I'll leave it as is unless you have any specific suggestions.

Reply:
Ah, I want `0` on cache miss for that metric. This way, you have from this counter both the number of hits (value of the counter) as well as the number of times we checked the cache (number of times the metric was emitted).

That's the tricky one to plumb in, otherwise your approach is the better option.